asaplib/cluster/ml_cluster_fit.py

```diff
@@ -100,115 +102,344 @@ def pack(self):
+        self.decision_graph = None
 
+        # Clusters must lie at a distance greater than self.delta_cut (float) apart to be designated as independent
+        # clusters
+        self.delta_cut = None
+
+        # If a "cluster" has a density smaller than self.dens_cut (float) it is discarded as noise
+        self.dens_cut = None
+
+
+        self.cluster = None  # numpy array of shape (N_data, )
+        self.centers = None  # numpy array of cluster centers of shape(num_centers, )
+        self.halo = None  # numpy array of halo points of shape (N_data, )
+
+    def get_dc(self, data):
         """
+        Compute the cutoff distance given the data.
 
         Parameters
         ----------
-        dmatrix: The distance matrix
-        rho: The log densities of the points
+        data: The data array of shape (Nele, proj_dim) where N is the number of data points and proj_dim is the number of
+        projected components of the kernel matrix.
 
         Returns
         -------
-        cluster label
+        self.dc, the cutoff distance
         """
 
-        if rho is None:
-            raise ValueError('for fdb it is better to compute kernel density first')
-
-        delta, nneigh = self.estimate_delta(dmatrix, rho)
-
-        # if there's no input values for rhomin and deltamin, we use simple heuristics
-        if self.rhomin < 0:
-            self.rhomin = 0.2 * np.mean(rho) + 0.8 * np.min(rho)
-        if self.deltamin < 0:
-            self.deltamin = np.mean(delta)
+        Nele = data.shape[0]
+
+        # The number of neighbours to consider is dictated by the percent parameter provided (default is 2% of the
+        # total data points.
+        n_neigh = int(self.percent/100.*Nele)
+
+        # Compute with 4 times the recommended percentage of neighbours in order that we may re-estimate the percent
+        # value supplied (rper) based on the data.
+        n_neigh_comp = int(4.*self.percent/100.*Nele)
+        neigh = NearestNeighbors(n_neighbors=n_neigh_comp).fit(data)
+        self.distances, self.indices = neigh.kneighbors(data)
+
+        # The cutoff distance dc is the average distance (averaged over all data points) of the "0.02*Nele"th neighbour
+        dc = np.average(self.distances[:, n_neigh])
+
+        # To store the local densities of each data point
+        dens = np.zeros(data.shape[0])
+        tt = 1.0
+        factor = 1.0
+
+        # We update dc based on a re-estimated percentage rper. Stop when rper is close to self.percent
+        while tt > 0.05:
+            dc = dc*factor
+            for i in range(Nele):
+                a = self.distances[i, :]
+                dens[i] = len(a[(a <= dc)])
+            rper = 100.*np.average(dens)/Nele
+            tt = rper/self.percent - 1.
+            if tt > 0.:
+                factor = factor/1.001
+            else:
+                factor = factor*1.001
+            tt = abs(tt)
+        self.dc = dc
+
+    def get_decision_graph(self, data):
```

Comment:
This method doesn't actually produce the decision graph. I think it would be good for me to add your plotting code to this method as well as a switch to turn the plotting function on and off in some sensible place.